### DISCUSSION TITLE
The top class can.bus optimize suggestion.

### DISCUSSION THREAD

**Discussion by HankXiao** [hardbyte/python-can#1728] [Ideas]
I really like the notifier, period sending task, and asyncio support. Additionally, I have some suggestions regarding can.Bus. Initially, I thought that can.Bus is a top-class solution that would allow me to forget about the hardware I am currently using. However, it appears to have some differences, particularly with PeakCan and ValueCan3 (neovi).

Peak devices can obtain a handle for each channel within the same hardware, whereas neovi devices can only acquire one handle for each device, regardless of the number of channels it has.

In this case, I need to open the neovi device during the can.Bus initialization. However, I discovered that all channels will have the same settings (can or canfd, bitrate, etc.), and there is no way to configure different settings for each channel. Furthermore, I have not found any method to configure timing settings for neovi devices.

My project requires controlling multiple types of devices, and there may be virtual channels that correspond to real physical channels belonging to different devices. Therefore, I suggest that this project should consider abstracting a general device open and close method that can adapt to all devices if possible.

By the way, I noticed that when I use PeakCan with system timespan enabled, the timespan becomes inaccurate after the device has been connected for many hours. It might be necessary to calibrate the timespan during initialization and periodically, perhaps every minute, to ensure its accuracy.

**Comment by grant-allan-ctct** [comment]
This relates to only one of your interesting questions, but I wonder whether you can create a separate instance of can.Bus for each ValueCAN channel in order to end up with one handle per channel as you would like?  (Potentially you might need to turn on [Enable Server](https://cdn.intrepidcs.net/guides/neat/INCLUDE_The-neoVI-Explorer-Configuration-Utility.html#system-settings) for your device to get this to work, or maybe it is not even possible - just mentioning it as an idea.)

**Reply by HankXiao** [reply]
After turn on Enable Server, now it is work with multiple handle, thank you.

**Comment by pierreluctg** [comment]
Hi @HankXiao, regarding the neovi interface, what OS are you using? The solution to your issue may be different depending on the OS.

@grant-allan-ctct is correct, if you are on Windows, you do want to enable the server mode in order to be able to create multiple handle to the same device.

**Reply by HankXiao** [reply]
My project is on windows 10/11, thanks for information, it is solved.
